File: src/main/java/leetcode/topinterview/LeetCode169.java
package leetcode.topinterview;

/**
 * @author wujingxinit@outlook.com
 * @date 2022/9/27 13:02
 */
public class LeetCode169 {

    //study
    //https://www.youtube.com/watch?v=2s7b0zs4Vf4
    public int majorityElement(int[] nums) {
        int count = 0;
        int res = 0;
        for (int num : nums) {
            if (count == 0) {
                res = num;
            }

            if (num != res) {
                count--;
            } else {
                count++;
            }
        }

        return res;
    }
}

File: go/src/leetcode/hot100/LeetCode169.go
package hot100

func majorityElement(nums []int) int {
	if len(nums) == 0 {
		return 0
	}
	mp := make(map[int]int)
	for _, n := range nums {
		mp[n] = mp[n] + 1
	}
	for k, v := range mp {
		if v > len(nums)/2 {
			return k
		}
	}
	return -1
}
